src/OpenLoco/src/Map/MapGenerator.cpp

```diff
@@ -648,6 +660,154 @@ namespace OpenLoco::World::MapGenerator
+        }
+    };
+
+    class PngOps
+    {
+    public:
+        static std::unique_ptr<PngImage> loadPng(std::string filename)
+        {
+            std::ifstream inFile(filename, std::ios::binary);
+
+            png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+            if (!png)
+            {
+                Logging::error("Failed to create PNG read struct");
+                return nullptr;
+            }
+
+            png_infop info = png_create_info_struct(png);
+            if (!info)
+            {
+                Logging::error("Failed to create PNG info struct");
+                png_destroy_read_struct(&png, NULL, NULL);
+                return nullptr;
+            }
+
+            // Set error handler
+            if (setjmp(png_jmpbuf(png)))
+            {
+                png_destroy_read_struct(&png, &info, NULL);
+                Logging::error("Failed to set png error handler");
+                return nullptr;
+            }
+
+            png_set_read_fn(png, static_cast<void*>(&inFile), [](png_structp png_ptr, png_bytep data, png_size_t length) {
+                std::istream* inStream = static_cast<std::istream*>(png_get_io_ptr(png_ptr));
+                inStream->read(reinterpret_cast<char*>(data), length);
+            });
+
+            png_read_info(png, info);
+
+            int width = png_get_image_width(png, info);
+            int height = png_get_image_height(png, info);
+            // png_byte colorType = png_get_color_type(png, info);
+
+            auto pngImage = std::make_unique<PngImage>(width, height);
+
+            png_bytep* rowPointers = new png_bytep[height];
+            for (int y = 0; y < height; y++)
+            {
+                rowPointers[y] = &pngImage->imageData[y * width * 4];
+            }
+            png_read_image(png, rowPointers);
+
+            // cleanup image
+            delete[] rowPointers;
+            png_destroy_read_struct(&png, &info, NULL);
+            inFile.close();
+
+            return pngImage;
+        }
+
+        /*
+        static void savePng()
+        {
+        }
+        */
+    };
+
+    class HeightMapTerrainGenerator
+    {
+    public:
+        void generate(HeightMapRange heightMap)
+        {
+            generateFromHeightmapPng(heightMap);
+        }
+
+    private:
+        void openUiPngBrowser()
+        {
+            auto basePath = Platform::getUserDirectory();
+            strncpy(&_savePath[0], basePath.string().c_str(), std::size(_savePath));
+
+            GameCommands::LoadSaveQuitGameArgs args{};
+            args.option1 = GameCommands::LoadSaveQuitGameArgs::Options::dontSave;
+            args.option2 = LoadOrQuitMode::loadGamePrompt;
+            GameCommands::doCommand(args, GameCommands::Flags::apply);
+        }
+
+        int generateFromHeightmapPng(HeightMapRange heightMap)
+        {
+            openUiPngBrowser(); // seems like game commands like this cannot be run here
```

Comment:
help needed here - file browser will not display no matter what i try. i presume its because we're in some pseudo update loop in MapGenerator::generate()